### PR TITLE
refactor(storage/fs): adjust the declarative storage abstractions

### PR DESCRIPTION
**Pull request by GeorgeMac**
Fixes FLI-747

I will hold off on doing this until #2538 is delivered and this has been updated to include that work.
To avoid causing any disruption there and so @erka can get his work in 👍 .

This change is going to mix up the abstractions a little.
The motivation is this will help with some work coming up, where a couple of the backends will actually support multiple versions of snapshots (keyed by their git tag / revision or oci tag / digest). Clients will be able to request snapshots by reference or potentiall even revision.

The refactor drops the idea of `fs.SnaphotSource` altogether.
Instead, it introduces the concept of `fs.SnapshotStore`:
```go
// SnapshotStore is a type which has a single function View.
// View is a functional transaction interface for reading a snapshot
// during the lifetime of a supplied function.
type SnapshotStore interface {
	// View accepts a function which takes a *StoreSnapshot.
	// The SnapshotStore will supply a snapshot which is valid
	// for the lifetime of the provided function call.
	View(func(storage.ReadOnlyStore) error) error
	fmt.Stringer
}
```

This new store exposes a kind of funtional transaction interface (similar to e.g. boltdb).
Instead of making the top-level `fs.Store` managed locking and handle subscribing to changes, it instead pushes the synchronization responsibilities down onto the snapshot stores themselves (previously called sources).
Now the individual stores have to take care of managing how and when to update their snapshots.

To achieve this, the new abstraction exposes a function `View()` which itself takes a function.
The function provided to View is given the snapshot instead.
It is on the store implementers to call the provided function inside a read lock.
The write lock can and is held by the store when changing the snapshots themselves.

This allows the top-level `fs.Store` to define what happens per read-lock, without having to own the lock itself.
In subsequent change, the `View()` function will be updated to take a target _reference_.
This will allow the various store implementations to support managing multiple snapshots for different references.
Whether or not a store supports arbitrary references, will be up to the store (initially Git and OCI will).

### CODE REVIEW

**Comment by erka** [comment]
@GeorgeMac please don't wait for me. It looks like it will take some time to finish Azure blob storage

**Comment by GeorgeMac** [comment]
Cheers for the heads up @erka 🙏 If I do end up getting this in before, I am happy to point out how the interfaces have changed. So you can adjust accordingly. I can update the open issue too with more up to date details.

File: internal/cmd/grpc.go [hunk 1]
```diff
@@ -154,119 +144,12 @@ func NewGRPCServer(
...
-			storageoci.WithPollInterval(cfg.Storage.OCI.PollInterval),
-		)
-		if err != nil {
-			return nil, err
-		}
-
-		store, err = fs.NewStore(logger, source)
+	default:
+		// otherwise, attempt to configure a declarative backend store
+		store, err = fsstore.NewStore(ctx, logger, cfg)
```
[the remark this answers is not in the document]
Comment (GeorgeMac) [reply internal/cmd/grpc.go:149]: Just don't open that new package and it's like all that code doesn't exist.

**Review by markphelps — APPROVED** [review]
this is amazing! love these abstraction changes!!

**Comment by markphelps** [comment]
@erka let me know if I can help unblock in any way with the Azure blob store changes! Thank you for taking that on

**Comment by erka** [comment]
@markphelps this PR doesn't impact my work much. The challenge for me is Aruzite, azure-sdk-for-go and auth between them for testing. I also have never used `dagger` and have white spots.

**Comment by markphelps** [comment]
> @markphelps this PR doesn't impact my work much. The challenge for me is Aruzite, azure-sdk-for-go and auth between them for testing. I also have never used `dagger` and have white spots.

gotcha. I'm happy to help with getting the Dagger tests working as well with Azurite (never used Azurite but looks cool)

**Pull request merged**